DESIGN.md

```diff
@@ -0,0 +1,364 @@
+Pirouette's Design
+==================
+
+This document aims at giving an overview of the crucial parts of Pirouette, and
+pointers to get familiar with the code. As such, it contains three parts.
+
+1. The [first part][languages] describe how Pirouette supports several input
+   languages as well as two languages that are particularly relevant to us: [the
+   example language] and the support for Plutus IR – [Plutus intermediate
+   representation].
+
+2. The [second part][transformations] describes the transformations that
+   Pirouette performs on the input language to get it in a shape ready for
+   symbolic execution.
+
+3. The [third part][symbolic engine] digs into what constitutes the heart of
+   Pirouette: its symbolic engine, and in particular the [symbolic evaluator]
+   and the [symbolic prover].
+
+Languages
+---------
+
+[languages]: #languages
+
+### Pirouette's genericity
+
+Pirouette attempts to be generic in the input language by only assuming that it
+is System F-like. Anyone can then come and define their instance of
+`LanguageBuiltins`, which is Pirouette's way of defining the constants, builtin
+terms and builtin types of a language. The definition of the classes can be
+found in [`Pirouette.Term.Syntax.Base`]. For a simple, concrete instantiation of
```

Comment:
```suggestion
found in [`Pirouette.Term.Syntax.Base`]. For a simple instantiation of
```

Reply:
What would be a non-concrete instantiation?

Reply:
This question makes me realise how misleading that word can be. I just meant something we could look at in the current repository. I'll take your suggestion to remove the word altogether.